mantidproject/mantid
# Add calibration as pre-processing step to ISIS Reflectometry algorithms

**Description of work.**

Calls the new `ReflectometryISISCalibration` algorithm as part of `ReflectometryISISPreprocess`, if a path to a calibration file has been supplied.

Also refactors `ReflectometryISISLoadAndProcess` so that it can pass a calibration file path to the pre-processing step. Note that this algorithm only calls `ReflectometryISISPreprocess` (and therefore applies calibration) if the workspace cannot be loaded from the ADS. **Is this OK or should we also be calibrating workspaces that have been loaded from the ADS?**

The calibration file property is also exposed by `ReflectometryReductionOneLiveData` so that it can be passed by that algorithm to `ReflectometryISISLoadAndProcess`.

This PR also updates `ReflectometryISISCalibration` so that it provides unique names for the calibration table workspaces that it outputs, and to allow it to handle workspace groups.

**To test:**

- Check that a normal reduction run via the interface (investigation ID `1120015` and cycle `11_3` could be used for this) completes as it would have before these changes i.e. there should be no `Calib_Table_{run number}` workspace in the ADS and no call to `ReflectometryISISCalibration` in the histories of the output workspaces.
- Check that the preview tab still works as it did before these changes (i.e. when data is loaded and regions selected the reduction completes successfully).
- Run `ReflectometryISISLoadAndProcess` with the following inputs:

InputRunList - INTER61135
ThetaIn - 0.5
CalibrationFile - load `pixel_centers_repeat.dat` from the example data on the issue
AnalysisMode - MultiDetectorAnalysis
WavelengthMin - 2
WavelengthMax - 5
IOMonitorIndex - 1
MomentumTransferStep - 0.02

- Check `Calib_Table_{run number}` workspace in ADS and that `Detector Y Coordinate` column is different to the Y value (the second list value) in the `Detector Position` column. Check the histories of `IvsQ_61135` and `IvsQ_binned_61135` have a call to `ReflectometryISISCalibration`.

Fixes #32022.

*This does not require release notes* because they will be added in a future PR that allows a calibration file to be passed in via the GUI (see issue #34111).



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
